completions/tmux

```diff
@@ -0,0 +1,297 @@
+# If BASH_COMPLETION_DEBUG is set, it will be printed to stderr.
+# When running with `set -x`, the _comp_cmd_tmux__log call itself will be
+# printed.
+#
+# @param $1 Message to log
+_comp_cmd_tmux__log()
+{
+    if [[ ${BASH_COMPLETION_DEBUG-} ]]; then
+        printf 'tmux bash completion: %s\n' "$1" >&2
+    fi
+}
+
+# Parse usage output from tmux.
+#
+# @param $1 Usage from tmux, not including the (sub)command name.
+# @var[out] options associative array mapping options to their value types, or
+#     to the empty string if the option doesn't take a value
+# @var[out] args indexed array of positional arg types
+_comp_cmd_tmux__parse_usage()
+{
+    options=()
+    args=()
+
+    local i j
+    local words
+    _comp_split words "$1"
+    for ((i = 0; i < ${#words[@]}; i++)); do
+        case "${words[i]}" in
+            "[-"*"]")
+                # One or more options that don't take arguments, either of the
+                # form `[-abc]` or `[-a|-b|-c]`
+                for ((j = 2; j < ${#words[i]} - 1; j++)); do
+                    if [[ ${words[i]:j:1} != [-\|] ]]; then
+                        options+=(["-${words[i]:j:1}"]="")
+                    fi
+                done
+                ;;
+            "[-"*)
+                # One option that does take an argument.
+                if [[ ${words[i + 1]} != *"]" ]]; then
+                    _comp_cmd_tmux__log \
+                        "Can't parse option: '${words[*]:i:2}' in '$1'"
+                    break
+                fi
+                options+=(["${words[i]#"["}"]="${words[i + 1]%"]"}")
+                ((i++))
+                ;;
+            -*)
+                _comp_cmd_tmux__log "Can't parse option '${words[i]}' in '$1'"
+                break
+                ;;
+            *)
+                # Start of positional arguments.
+                args=("${words[@]:i}")
+                break
+                ;;
+        esac
+    done
+
+    local arg
+    for arg in "${!options[@]}"; do
+        _comp_cmd_tmux__log "option: ${arg} ${options["$arg"]}"
+    done
+    for arg in "${args[@]}"; do
+        _comp_cmd_tmux__log "arg: ${arg}"
+    done
+}
+
+# Complete a value either as the argument to an option or as a positional arg.
+#
+# @param $1 subcommand that the value is for, or 'tmux' if it's top-level
+# @param $2 type of the value, from _comp_cmd_tmux__parse_usage()
+_comp_cmd_tmux__value()
+{
+    local subcommand="$1" option_type="$2"
+    _comp_cmd_tmux__log \
+        "Trying to complete '$option_type' for subcommand '$subcommand'"
+
+    # To get a list of these argument types, look at `tmux -h` and:
+    #
+    # tmux list-commands -F "#{command_list_usage}" |
+    #   sed 's/[][ ]/\n/g' |
+    #   grep -v ^- |
+    #   sort -u
+    #
+    # TODO: Complete more option types.
+    case "$option_type" in
+        command)
+            _comp_compgen_split -l -- \
+                "$(LC_ALL=C tmux list-commands -F "#{command_list_name}")"
```

Comment:
Would be good to pass around the original tmux command invoked instead of hardcoding just `tmux` from `$PATH` (or as an alias or a function) -- `tmux` might not be in `$PATH` (or the like) in the first place, or the `tmux` there might not be the one that was invoked (maybe another one with a full path to it was).

Reply:
To add more context, one simple way is to use `"$1"` (at the top-level function) or `"${comp_words[0]}"` (in the helper functions that are called from the top-level function). A more robust way is to use `_comp_dequote "${comp_words[0]}"` as `_comp_compgen_help__get_help_lines` is doing:

https://github.com/scop/bash-completion/blob/90162b0a6cb31fe32acfe6af5079c9c4454dc8fb/bash_completion#L1561-L1569

Note: I try to slightly change the behavior of `_comp_dequote` in https://github.com/scop/bash-completion/pull/1357/commits/3e088b7a8e13a7c352eaad8b17c766baf35a680c (PR #1357), so that one does not need to prefix `|| REPLY=${comp_words[0]-}`.

Reply:
After checking the implementation, I think maybe we can just use this function `_comp_compgen_help__get_help_lines` for broader purposes (with a small adjustment). Even with the current implementation, we can use `_comp_compgen_help__get_help_lines` for the present purpose in the following way:

```bash
_comp_compgen_help__get_help_lines -- list-commands -F "#{command_list_name}" &&
  REPLY=("${_lines[@]}") &&
  _comp_compgen -- -W '"${REPLY[@]}"'
```

Then, we may consider renaming this function to `_comp_run_cmd` and change its output variables from `_lines` to `REPLY`.

Reply:
`_comp_compgen_help__get_help_lines` is private, so I shouldn't use it as is, right?

Should I wait for changes to `_comp_compgen_help__get_help_lines` and/or `_comp_dequote`, or use `"${comp_args[0]}"` / `"$1"` for now?

Reply:
> `_comp_compgen_help__get_help_lines` is private, so I shouldn't use it as is, right?

Yes, so I wrote

> > Then, we may consider renaming this function to `_comp_run_cmd` and change its output variable from `_lines` to `REPLY`.

----

> Should I wait for changes to `_comp_compgen_help__get_help_lines` and/or `_comp_dequote`, or use `"${comp_args[0]}"` / `"$1"` for now?

You can change `_comp_compgen_help__get_help_lines` if other maintainers agree with it.

Reply:
I added `_comp_cmd_tmux__run` to do this without `_comp_compgen_help__get_help_lines` for now. Happy to switch to that later if maintainers want to rename it and make it public.

> `_comp_dequote "${comp_args[0]-}" || REPLY=${comp_args[0]-}`

`_comp_dequote`'s documentation says that `REPLY` is an array, so I didn't use that exactly. Should `_comp_compgen_help__get_help_lines` be changed too?

I also updated the semicolon detection to treat `REPLY` as an array too.

Reply:
OK, we can update `_comp_compgen_help__get_help_lines` (`_comp_run_cmd`) in another PR.

Reply:
> Should `_comp_compgen_help__get_help_lines` be changed too?

Yeah, and it is actually addressed in https://github.com/scop/bash-completion/commit/3e088b7a8e13a7c352eaad8b17c766baf35a680c in #1357 (where `REPLY` is assigned after `REPLY=()`). That was actually [a fix](https://github.com/scop/bash-completion/compare/372d4992eebecb338ab1b5fdd78bdbd850b91217..ebe65544ee9814985d4dd92bce3c24942787f81f) after [the CI \#2838 failure](https://github.com/scop/bash-completion/actions/runs/14282501108/job/40033674783).